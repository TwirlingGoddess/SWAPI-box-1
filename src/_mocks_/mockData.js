const mockVehicleApiResponse = {
  count: 39,
  next: 'https://swapi.co/api/vehicles/?page=2',
  previous: null,
  results: [
    {
      cargo_capacity: '50000',
      consumables: '2 months',
      cost_in_credits: '150000',
      created: '2014-12-10T15:36:25.724000Z',
      crew: '46',
      edited: '2014-12-22T18:21:15.523587Z',
      films: ['https://swapi.co/api/films/5/', 
        'https://swapi.co/api/films/1/'],
      length: '36.8',
      manufacturer: 'Corellia Mining Corporation',
      max_atmosphering_speed: '30',
      model: 'Digger Crawler',
      name: 'Sand Crawler',
      passengers: '30',
      pilots: [],
      url: 'https://swapi.co/api/vehicles/4/',
      vehicle_class: 'wheeled'
    }
  ]
};

const mockPeopleApiResponse = {
  count: 87,
  next: 'https://swapi.co/api/people/?page=2',
  previous: null,
  results: [
    {
      birth_year: '19BBY',
      created: '2014-12-09T13:50:51.644000Z',
      edited: '2014-12-20T21:17:56.891000Z',
      eye_color: 'blue',
      films: ['https://swapi.co/api/films/2/',
        'https://swapi.co/api/films/6/',
        'https://swapi.co/api/films/3/'],
      gender: 'male',
      hair_color: 'blond',
      height: '172',
      homeworld: 'https://swapi.co/api/planets/1/',
      mass: '77',
      name: 'Luke Skywalker',
      skin_color: 'fair',
      species: ['https://swapi.co/api/species/1/'],
      starships: ['https://swapi.co/api/starships/12/',
        'https://swapi.co/api/starships/22/'],
      url: 'https://swapi.co/api/people/1/',
      vehicles: ['https://swapi.co/api/vehicles/14/',
        'https://swapi.co/api/vehicles/30/']
    }
  ]
};

const mockParseData = [
  {
    birth_year: '19BBY',
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    eye_color: 'blue',
    films: ['https://swapi.co/api/films/2/',
      'https://swapi.co/api/films/6/',
      'https://swapi.co/api/films/3/'],
    gender: 'male',
    hair_color: 'blond',
    height: '172',
    homeworld: 'https://swapi.co/api/planets/1/',
    mass: '77',
    name: 'Luke Skywalker',
    skin_color: 'fair',
    species: ['https://swapi.co/api/species/1/'],
    starships: ['https://swapi.co/api/starships/12/', 
      'https://swapi.co/api/starships/22/'],
    url: 'https://swapi.co/api/people/1/',
    vehicles: ['https://swapi.co/api/vehicles/14/', 
      'https://swapi.co/api/vehicles/30/']
  }
];

const mockPlanetApiResponse = {
  count: 61,
  next: 'https://swapi.co/api/planets/?page=2',
  previous: null,
  results: [
    {
      climate: 'temperate',
      created: '2014-12-10T11:35:48.479000Z',
      diameter: '12500',
      edited: '2014-12-20T20:58:18.420000Z',
      films: ['https://swapi.co/api/films/6/', 
        'https://swapi.co/api/films/1/'],
      gravity: '1 standard',
      name: 'Alderaan',
      orbital_period: '364',
      population: '2000000000',
      residents: ['https://swapi.co/api/people/5/',
        'https://swapi.co/api/people/68/',
        'https://swapi.co/api/people/81/'],
      rotation_period: '24',
      surface_water: '40',
      terrain: 'grasslands, mountains',
      url: 'https://swapi.co/api/planets/2/'
    }
  ]
};

const mockPeopleObject = [{'Homeworld': undefined, 
  'Name': 'Luke Skywalker', 
  'Population': undefined, 
  'Specie': undefined, 'id': 
'people0', 
  'keyList': 'people'
}];

const mockPeopleObject2 = {
  'Homeworld': undefined, 
  'Name': 'Leia', 
  'Population': undefined, 
  'Specie': undefined, 'id': 
'people2', 
  'keyList': 'people'
};

const mockFavorites = [
  {'Homeworld': undefined, 
    'Name': 'Luke Skywalker', 
    'Population': undefined, 
    'Specie': undefined, 'id': 
'people0', 
    'keyList': 'people'
  },
  {
    'Homeworld': undefined, 
    'Name': 'Leia', 
    'Population': undefined, 
    'Specie': undefined, 'id': 
  'people2', 
    'keyList': 'people'
  }
];

const filmCrawlResponse = [{
  crawl: 'Film Crawl',
  title: 'Movie Title',
  date: '1982-09-18',
  opening_crawl: 'There is unrest in the Galactic'},
{
  crawl: 'Film Crawl',
  title: 'Movie Title',
  date: '1982-09-18',
  opening_crawl: 'There is unrest in the Galactic'
}];


const filmCrawl = {
  crawl: 'Film Crawl',
  title: 'Movie Title',
  date: '1982-09-18',
  opening_crawl: 'There is unrest in the Galactic'
};

const vehicleData = [{
  Class: 'sail barge',
  id: 'vehicles0',
  Model: 'Modified Luxury Sail Barge',
  Name: 'Sail barge',
  Passengers: '500',
  keyList: 'vehicles'
}];

const planetData = [{
  Climate: 'temperate',
  id: 'planets0',
  Name: 'Alderaan',
  Residents: 'Alderaan',
  Population: '2000000000',
  Terrain: 'grasslands, mountains',
  keyList: 'planets'
}];

const parsedVehicleData = [
  {
    'name': 'Sail barge',
    'model': 'Modified Luxury Sail Barge',
    'manufacturer': 'Ubrikkian Industries Custom Vehicle Division',
    'cost_in_credits': '285000',
    'length': '30',
    'max_atmosphering_speed': '100',
    'crew': '26',
    'passengers': '500',
    'cargo_capacity': '2000000',
    'consumables': 'Live food tanks',
    'vehicle_class': 'sail barge',
    'pilots': [],
    'films': [
      'https://swapi.co/api/films/3/'
    ],
    'created': '2014-12-18T10:44:14.217000Z',
    'edited': '2014-12-22T18:21:15.807906Z',
    'url': 'https://swapi.co/api/vehicles/24/'
  }
];

const parsedPlanetData = [
  {
    climate: 'temperate',
    created: '2014-12-10T11:35:48.479000Z',
    diameter: '12500',
    edited: '2014-12-20T20:58:18.420000Z',
    films: ['https://swapi.co/api/films/6/', 
      'https://swapi.co/api/films/1/'],
    gravity: '1 standard',
    name: 'Alderaan',
    orbital_period: '364',
    population: '2000000000',
    residents: ['https://swapi.co/api/people/5/',
      'https://swapi.co/api/people/68/',
      'https://swapi.co/api/people/81/'],
    rotation_period: '24',
    surface_water: '40',
    terrain: 'grasslands, mountains',
    url: 'https://swapi.co/api/planets/2/'
  }
];

export { mockPlanetApiResponse, 
  mockVehicleApiResponse, 
  mockPeopleApiResponse, 
  mockPeopleObject, 
  mockPeopleObject2,
  mockFavorites,
  filmCrawl,
  filmCrawlResponse,
  mockParseData,
  vehicleData,
  planetData,
  parsedVehicleData,
  parsedPlanetData
};

