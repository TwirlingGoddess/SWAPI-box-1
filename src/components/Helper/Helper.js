import React from 'react';

export const cleanDataFunc = (dataObject) => {
  if (dataObject.next.includes('people')) {
    return peopleObject(dataObject.results);
  }
  if (dataObject.next.includes('planets')) {
    return planetObject(dataObject.results);
  }
  if (dataObject.next.includes('vehicles')) {
    return vehicleObject(dataObject.results);
  }
}

export const peopleObject = async (parsedData) => {
  const unresolvedPromises = parsedData.map( async(person,index) => {
    const keyList = 'people';
    const homeworld = await nestedFetch(person.homeworld);
    const homeworldName = homeworld.name;
    const homeworldPopulation = homeworld.population;
    const specieObj = await nestedFetch(person.species);
    const specie = specieObj.name;
    const name = person.name;
    const id = keyList + index;
    return {id, keyList, Homeworld: homeworldName, Population: homeworldPopulation, Specie: specie, Name:name};
  });
  return Promise.all(unresolvedPromises);
}


export const planetObject = async (parsedData) => {
  const unresolvedPromises = parsedData.map(async(planet,index) => {
    const keyList = 'planets';  
    const planetName = planet.name;
    const planetTerrain = planet.terrain;
    const planetPopulation = planet.population;
    const planetClimate = planet.climate;
    const residents = await residentsFetch(planet.residents);
    const id = keyList + index;
    return {id, keyList, Name: planetName, Terrain: planetTerrain, Population: planetPopulation, Climate: planetClimate, Residents: residents};
  });
  return Promise.all(unresolvedPromises);
}

export const vehicleObject = (parsedData) => {
  const vehicleArray = parsedData.map((vehicle, index) => {
    const keyList = 'vehicles'; 
    const vehicleName = vehicle.name;
    const vehicleModel = vehicle.model;
    const vehicleClass = vehicle.vehicle_class;
    const numberOfPassengers = vehicle.passengers;
    const id = keyList + index;
    return {id, keyList, Name: vehicleName, Model: vehicleModel, Class: vehicleClass, Passengers:numberOfPassengers};
  });
  return vehicleArray;
}

export const nestedFetch = async (url) => {
  const fetchURL= await fetch(url);
  const parseObject= await fetchURL.json();
  return parseObject;
}

export const residentsFetch = async (residentsUrls) => {
  const unresolvedPromises = residentsUrls.map(async(url) => {
    const residentURLfetch = await fetch(url);
    const parseResidents = await residentURLfetch.json();
    return parseResidents.name;
  });
  return Promise.all(unresolvedPromises);
}

export const makeApiCall = async (category) => {
  const url = `https://swapi.co/api/${category}`;
  return apiCall(url)
}

 export const apiCall = async (url) => {
  const fetchURL = await fetch(url);
  const parseObject = await fetchURL.json();
  const cleanData = await cleanDataFunc(parseObject);
  sendToLocalStorage(url, cleanData);
  return cleanData;
 }

export const sendToLocalStorage = (key, selectedData) => {
  localStorage.setItem(key, JSON.stringify(selectedData));
}

export const getFromLocalStorage = (key) => {
  return JSON.parse(localStorage.getItem(key));
}
