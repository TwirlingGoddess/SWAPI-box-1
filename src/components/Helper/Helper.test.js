import React from 'react';
import { makeApiCall, cleanDataFunc } from './Helper';
import { shallow } from 'enzyme';
import { fetchedPeopleData } from '../../_mocks_/mockData';

jest.autoMockOn()

  describe ('makeApiCall', () => {

  beforeEach(()=> { 
    jest.resetAllMocks();
    window.fetch = jest.fn().mockImplementation(() => {
      Promise.resolve({ json: () => Promise.resolve({
        data: 'mock'
      })
      })
    })
  });

  it('should match snapshot', async () => {
    await expect(makeApiCall).toMatchSnapshot()
  })

  it('calls fetch with the correct data', async () => {
    const expectedFetchBody = {
      method: 'GET',
      body: JSON.stringify({ result: fetchedPeopleData }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    const url = 'https://swapi.co/api/people'
    const fetchURL = window.fetch(url) 
    expect(window.fetch).toHaveBeenCalled();
  });

  it('calls cleandata', () => {
    
  });

  it('calls saveToLocalStorage', () => {

  });

  it('saves to local storage', () => {
    
  });

  it('returns clean data', () => {
    
  });

  describe('cleanData', () => {

    it('returns an object with the correct information', () => {
    
    });
  });

  describe('peopleObject', () => {

    it('returns a people object', () => {
    
    });
  });

  describe('planetObject', () => {

    it('returns a planet object', () => {
    
    });
  });

  describe('nestedFetch', () => {

    it('returns an object with the correct information', () => {
    
    });
  });

  describe('residentsFetch', () => {

    it('returns an object with the correct information', () => {
    
    });
  });

  describe('sendToLocalStorage', () => {

    it('saves to local storage', () => {
    
    });
  });

  describe('getFromLocalStorage', () => {

    it('retrieves info from local storage ', () => {
    
    });
  });
});