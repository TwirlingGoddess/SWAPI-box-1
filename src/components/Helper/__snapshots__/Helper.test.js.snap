// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeApiCall should match snapshot 1`] = `[Function]`;
